pkg/util/kubernetes/kubelet/podwatcher.go

```diff
@@ -103,51 +91,60 @@ func (w *PodWatcher) computeChanges(podList []*Pod) ([]*Pod, error) {
 		isPodReady := IsPodReady(pod)
 
 		for _, container := range pod.Status.GetAllContainers() {
-			// We don't check container readiness as init containers are never ready
-			// We check if the container has an ID instead (has run or is running)
-			if !container.IsPending() {
-				// new container are always sent ignoring the pod state
-				if _, found := w.lastSeen[container.ID]; !found {
-					updatedContainer = true
-				}
-				w.lastSeen[container.ID] = now
-
-				// for existing ones we look at the readiness state
-				if _, found := w.lastSeenReady[container.ID]; !found && isPodReady {
-					// the pod has never been seen ready or was removed when
-					// reaching the unreadinessTimeout
-					updatedContainer = true
-				}
-
-				// update the readiness expiry cache
-				if isPodReady {
-					w.lastSeenReady[container.ID] = now
-				}
+			if container.IsPending() {
+				// We don't check container readiness as init
+				// containers are never ready. We check if the
+				// container has an ID instead (has run or is
+				// running)
+				continue
+			}
+
+			// new container are always sent ignoring the pod state
+			if _, found := w.lastSeen[container.ID]; !found {
+				updatedContainer = true
+			}
+			w.lastSeen[container.ID] = now
+
+			// for existing containers, check whether the
+			// readiness has changed since last time
+			if oldReadiness, found := w.oldReadiness[container.ID]; !found || oldReadiness != isPodReady {
+				// the pod has never been seen ready or was removed when
+				// reaching the unreadinessTimeout
+				updatedContainer = true
+			}
+
+			w.oldReadiness[container.ID] = isPodReady
+
+			if isPodReady {
+				w.lastSeenReady[container.ID] = now
 			}
 		}
 
-		// Detect changes in labels and annotations values
 		newLabelsOrAnnotations := false
-		// Detect changes in the pod phase
 		newPhase := false
-		if w.isWatchingTags() {
-			newTagsDigest := digestPodMeta(pod.Metadata)
-			if foundPod && newTagsDigest != w.tagsDigest[podEntity] {
-				// update tags digest
+		newTagsDigest := digestPodMeta(pod.Metadata)
+
+		// if the pod already existed, check whether tagsDigest or
+		// phase changed
+		if foundPod {
```

Comment:
should we also check the readiness of the pod as changed?

Reply:
It's actually tracked on a per-container basis (we don't check the actual container readiness, just the pod's, for reasons I didn't look too deeply into).

Reply:
We don't lose it, this logic is just for selecting which pods have changed. We pass the entire pod object if we detect that it has changed. But indeed, I agree it's confusing and warrants some refactoring 👍

Reply:
What I was referencing it is this line: https://github.com/DataDog/datadog-agent/pull/9680/files#diff-f1c9b074e4a34c38e5453e8658902be4d1a36e6bbba5024a6c6df3f8cafdd576R117 where we set `oldReadiness` of a container with the state of the pod `isPodReady`